consensus/src/dag/bootstrap.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright © Aptos Foundation
+
+use super::{
+    anchor_election::RoundRobinAnchorElection,
+    dag_driver::DagDriver,
+    dag_fetcher::{DagFetcher, FetchRequestHandler},
+    dag_handler::NetworkHandler,
+    dag_network::DAGNetworkSender,
+    dag_store::Dag,
+    order_rule::OrderRule,
+    rb_handler::NodeBroadcastHandler,
+    storage::DAGStorage,
+    types::DAGMessage,
+    CertifiedNode,
+};
+use crate::{
+    network::IncomingDAGRequest, state_replication::PayloadClient, util::time_service::TimeService,
+};
+use aptos_channels::{aptos_channel, message_queues::QueueStyle};
+use aptos_consensus_types::common::Author;
+use aptos_infallible::RwLock;
+use aptos_reliable_broadcast::{RBNetworkSender, ReliableBroadcast};
+use aptos_types::{
+    epoch_state::EpochState, ledger_info::LedgerInfo, validator_signer::ValidatorSigner,
+};
+use futures::stream::{AbortHandle, Abortable};
+use std::sync::Arc;
+use tokio_retry::strategy::ExponentialBackoff;
+
+pub fn bootstrap_dag(
+    self_peer: Author,
+    signer: ValidatorSigner,
+    epoch_state: Arc<EpochState>,
+    latest_ledger_info: LedgerInfo,
+    storage: Arc<dyn DAGStorage>,
+    rb_network_sender: Arc<dyn RBNetworkSender<DAGMessage>>,
+    dag_network_sender: Arc<dyn DAGNetworkSender>,
+    time_service: Arc<dyn TimeService>,
+    aptos_time_service: aptos_time_service::TimeService,
+    payload_client: Arc<dyn PayloadClient>,
+) -> (
+    AbortHandle,
+    AbortHandle,
+    aptos_channel::Sender<Author, IncomingDAGRequest>,
+    futures_channel::mpsc::UnboundedReceiver<Vec<Arc<CertifiedNode>>>,
+) {
+    let validators = epoch_state.verifier.get_ordered_account_addresses();
+    let current_round = latest_ledger_info.round();
+
+    let (ordered_nodes_tx, ordered_nodes_rx) = futures_channel::mpsc::unbounded();
+    let (dag_rpc_tx, dag_rpc_rx) = aptos_channel::new(QueueStyle::FIFO, 64, None);
+
+    // A backoff policy that starts at 100ms and doubles each iteration.
+    let rb_backoff_policy = ExponentialBackoff::from_millis(2).factor(50);
+    let rb = Arc::new(ReliableBroadcast::new(
+        validators.clone(),
+        rb_network_sender,
+        rb_backoff_policy,
+        aptos_time_service.clone(),
+    ));
+
+    let dag = Arc::new(RwLock::new(Dag::new(epoch_state.clone(), storage.clone())));
+
+    let anchor_election = Box::new(RoundRobinAnchorElection::new(validators));
+    let order_rule = OrderRule::new(
+        epoch_state.clone(),
+        latest_ledger_info,
+        dag.clone(),
+        anchor_election,
+        ordered_nodes_tx,
+    );
+
+    let (dag_fetcher, fetch_requester) = DagFetcher::new(
```

Comment:
I think dag_fetcher should also get order_rule, so it can call it every time a certificate is added to the dag (see my comments on the first PR).

Or actually, maybe it should be only part of Dag and called from add_node therein.

Reply:
I like the idea actually, but given how Zekun implemented oneshot receiver, I felt like that should be handled by the handler task and not the fetcher, because order rule can be a heavy operation imo.